library/psa_crypto.c

```diff
@@ -5393,12 +5486,15 @@ static psa_status_t psa_key_derivation_setup_kdf(
     psa_key_derivation_operation_t *operation,
     psa_algorithm_t kdf_alg )
 {
+#if !defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
+    (void)kdf_alg;
+#endif
     /* Make sure that operation->ctx is properly zero-initialised. (Macro
      * initialisers for this union leave some bytes unspecified.) */
     memset( &operation->ctx, 0, sizeof( operation->ctx ) );
 
     /* Make sure that kdf_alg is a supported key derivation algorithm. */
-#if defined(MBEDTLS_MD_C)
+#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
```

Comment:
Rather something like:
```c
#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)                                       
    if( PSA_ALG_IS_HKDF( kdf_alg ) )                                            
    {                                                                           
#endif /* MBEDTLS_PSA_BUILTIN_ALG_HKDF */                                       
#if defined(MBEDTLS_MD_C)                                                       
    if( PSA_ALG_IS_TLS12_PRF( kdf_alg ) ||                                      
        PSA_ALG_IS_TLS12_PSK_TO_MS( kdf_alg ) )                                 
    {                                                                           
#endif /* MBEDTLS_MD_C */                                                       
#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF) || defined(MBEDTLS_MD_C)              
        psa_algorithm_t hash_alg = PSA_ALG_HKDF_GET_HASH( kdf_alg );            
        size_t hash_size = PSA_HASH_SIZE( hash_alg );                           
        if( hash_size == 0 )                                                    
            return( PSA_ERROR_NOT_SUPPORTED );                                  
        if( ( PSA_ALG_IS_TLS12_PRF( kdf_alg ) ||                                
              PSA_ALG_IS_TLS12_PSK_TO_MS( kdf_alg ) ) &&                        
            ! ( hash_alg == PSA_ALG_SHA_256 || hash_alg == PSA_ALG_SHA_384 ) )  
        {                                                                       
            return( PSA_ERROR_NOT_SUPPORTED );                                  
        }                                                                       
        operation->capacity = 255 * hash_size;                                  
        return( PSA_SUCCESS );                                                  
#endif /* defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF) || defined(MBEDTLS_MD_C) */     
#if defined(MBEDTLS_MD_C)                                                       
    }                                                                           
#endif /* MBEDTLS_MD_C */                                                       
#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)                                       
    }                                                                           
#endif /* MBEDTLS_PSA_BUILTIN_ALG_HKDF */      
```

Reply:
I find that still messy. It would be cleaner with a dedicated function `bool is_kdf_supported(psa_algorithm_t kdf_alg)`.

Reply:
There is still room for improvement but I do think that this is already much better and readable than the current code.

Reply:
I thought about moving to a function as well, but opted to keep it simple for now to get this going.

Reply:
Simpler and smaller:
```
#if defined(MBEDTLS_PSA_BUILTIN_ALG_HKDF)
    if( PSA_ALG_IS_HKDF( kdf_alg ) )
        is_kdf_alg_supported = 1;
    else
#endif  
#if defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PRF)
    if( PSA_ALG_IS_TLS12_PRF( kdf_alg ) )
        is_kdf_alg_supported = 1;
    else
#endif
#if defined(MBEDTLS_PSA_BUILTIN_ALG_TLS12_PSK_TO_MS)
    if( PSA_ALG_IS_TLS12_PSK_TO_MS( kdf_alg ) )
        is_kdf_alg_supported = 1;
    else
#endif
        /* The KDF is unsupported */ ;
```

If none of the algorithms are enabled, the `if( is_kdf_alg_supported )` block is never executed. If the compiler doesn't notice, we're wasting a bit of code size. If the compiler does notice, it may complain about useless code (IIRC at least one of gcc or clang does so in similar cases, but I haven't checked whether it does in this particular case).

Mind you, if none of these KDF are supported, this whole function isn't needed. I think it would be worth adding an internal symbol `AT_LEAST_ONE_BUILTIN_KDF` (internal to `psa_crypto.c`, so no `MBEDTLS_` prefix, to avoid a complaint from `check-names.sh`) and guarding code accordingly.